gadget/install/install.go

```diff
@@ -377,6 +376,141 @@ func Run(model gadget.Model, gadgetRoot, kernelRoot, bootDevice string, options
+}
+
+// sysfsPathForBlockDevice returns the sysfs path for a block device.
+func sysfsPathForBlockDevice(device string) (string, error) {
+	partPath, err := os.Readlink(filepath.Join("/sys/class/block", device))
+	if err != nil {
+		return "", err
+	}
+	// Remove initial ../../ from partPath, and make path absolute
+	return filepath.Join("/sys/class/block", partPath), nil
+}
+
+// onDiskVolumeFromPartitionSysfsPath creates an OnDiskVolume that
+// matches the disk that contains the given partition sysfs path
+func onDiskVolumeFromPartitionSysfsPath(partPath string) (*gadget.OnDiskVolume, error) {
+	// Removing the last component will give us the disk path
+	diskPath := filepath.Dir(partPath)
+	disk, err := disks.DiskFromDevicePath(diskPath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot retrieve disk information for %q: %v", partPath, err)
+	}
+	onDiskVol, err := gadget.OnDiskVolumeFromDisk(disk)
+	if err != nil {
+		return nil, fmt.Errorf("cannot retrieve on disk volume for %q: %v", partPath, err)
+	}
+
+	return onDiskVol, nil
+}
+
+// applyLayoutToOnDiskStructure finds the on disk structure from a
+// sysfs partition path string and takes the laid out information from
+// laidOutVols and inserts it there.
+func applyLayoutToOnDiskStructure(onDiskVol *gadget.OnDiskVolume, partPath string, laidOutVols map[string]*gadget.LaidOutVolume, gadgetVolName string) (*gadget.OnDiskStructure, error) {
+	partDevice := filepath.Base(partPath)
+	onDiskStruct, err := structureFromPartDevice(onDiskVol, partDevice)
+	if err != nil {
+		return nil, fmt.Errorf("cannot find partition %q: %v", partPath, err)
+	}
+
+	laidOutStruct, err := laidOutStructureForDiskStructure(laidOutVols, gadgetVolName, onDiskStruct)
+	if err != nil {
+		return nil, err
+	}
+	// This fills LaidOutStructure, including (importantly) the ResolvedContent field
+	onDiskStruct.LaidOutStructure = *laidOutStruct
+
+	return onDiskStruct, nil
+}
+
+// WriteContent writes gadget content to the devices specified in
+// onVolumes. It returns the resolved on disk volumes.
+func WriteContent(onVolumes map[string]*gadget.Volume, observer gadget.ContentObserver,
+	gadgetRoot, kernelRoot string, model *asserts.Model, perfTimings timings.Measurer) ([]*gadget.OnDiskVolume, error) {
+
+	_, allLaidOutVols, err := gadget.LaidOutVolumesFromGadget(gadgetRoot, kernelRoot, model)
+	if err != nil {
+		return nil, fmt.Errorf("when writing content: cannot layout volumes: %v", err)
+	}
+
+	var onDiskVols []*gadget.OnDiskVolume
+	for volName, vol := range onVolumes {
+		var onDiskVol *gadget.OnDiskVolume
+		for _, volStruct := range vol.Structure {
+			// TODO fixme
+			if volStruct.Role == "mbr" {
+				continue
+			}
+			// TODO write raw content?
+			if volStruct.Filesystem == "" {
+				continue
+			}
+
+			// TODO maybe some changes will be needed when we have
+			// encrypted partitions, as the device won't be directly
+			// associated with a disk.
+			device := filepath.Base(volStruct.Device)
```

Comment:
seems like this Base could/should go inside the helper itself? anyway Base is idempotent

Reply:
Done